plugins/state_history_plugin/state_history_plugin.cpp

```diff
@@ -101,25 +101,27 @@ struct state_history_plugin_impl {
    void create_listener(const std::string& address) {
       const boost::posix_time::milliseconds accept_timeout(200);
       // connections set must only be modified by main thread; run listener on ship thread so sockets use default executor of the ship thread
-      fc::create_listener<Protocol>(thread_pool.get_executor(), _log, accept_timeout, address, "", [this](Protocol::socket&& socket) {
-         boost::asio::post(app().get_io_service(), [this, socket{std::move(socket)}]() mutable {
-            catch_and_log([this, &socket]() {
-               connections.emplace(new session(std::move(socket), boost::asio::make_strand(thread_pool.get_executor()), chain_plug->chain(),
-                                               trace_log, chain_state_log, finality_data_log,
-                                               [this](const chain::block_num_type block_num) {
-                                                  return get_block_id(block_num);
-                                               },
-                                               [this](const chain::block_id_type& block_id) {
-                                                  return chain_plug->chain().fetch_block_by_id(block_id);
-                                               },
-                                               [this](session_base* conn) {
-                                                  boost::asio::post(app().get_io_service(), [conn, this]() {
-                                                     connections.erase(connections.find(conn));
-                                                  });
-                                               }, _log));
+      fc::create_listener<Protocol>(thread_pool.get_executor(), _log, accept_timeout, address, "",
+         [this](const auto&) { return boost::asio::make_strand(thread_pool.get_executor()); },
+         [this](Protocol::socket&& socket) {
+            boost::asio::post(app().get_io_service(), [this, socket{std::move(socket)}]() mutable {
```

Comment:
Just for code conciseness, can we go back to placing the listener on main thread to avoid this post() entirely?

Reply:
Actually that doesn't work: https://github.com/AntelopeIO/spring/actions/runs/11238499086
We capture the `this` and we need the listener to shutdown when `thread_pool.stop()` is called in the `plugin_shutdown`. This can be cleaned up when we work: https://github.com/AntelopeIO/spring/issues/842

Reply:
> We capture the `this` and we need the listener to shutdown when `thread_pool.stop()` is called in the `plugin_shutdown`.

I'm not immediately seeing why that's a problem. `this` is still alive after `plugin_shutdown()` completes. I think something else we're missing. Maybe the already-created `strand` (that is using ship's thread pool's executor) expects to touch that executor when it's destroyed when `async_accept()` is torn down when the main thread's io_context is destroyed which is after the plugin is destroyed and thus the thread pool's executor.

But yeah can wait to touch this further later.